infra.lvm_tools/shrink_lv/roles/shrink_lv/files/shrink.sh

```diff
@@ -0,0 +1,281 @@
+    if [[ status -ne 0 ]]; then
+        echo "Failed to list block device properties for $2: $val" >&2
+        return 1
+    fi
+    if [[ "$val" != "lvm"  ]]; then
+        echo "Device $device_name is not of lvm type" >&2
+        return 1
+    fi
+    return 0
+}
+
+function parse_uuid() {
+    uuid=$(/usr/bin/awk '{print $1}'<<< "$1"|/usr/bin/awk -F'UUID=' '{print $2}')
+    val=$(/usr/bin/lsblk /dev/disk/by-uuid/"$uuid" -o NAME --noheadings 2>/dev/null)
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        echo "Failed to retrieve device name for UUID=$uuid" >&2
+        return $status
+    fi
+    echo "/dev/mapper/$val"
+    return 0
+}
+
+
+function shrink_volume() {
+    /usr/sbin/lvm lvreduce --resizefs -L "$2b" "$1"
+    return $?
+}
+
+
+function check_volume_size() {
+    current_size=$(get_current_volume_size "$1")
+    if [[ $current_size -lt $2 ]];then
+        echo "Current volume size for device $1 ($current_size bytes) is lower to expected $2 bytes" >&2
+        return 1
+    fi
+    if [[ $current_size -eq $2 ]]; then
+        echo "Current volume size for device $1 already equals $2 bytes" >&2
+        return 1
+    fi
+    return $?
+}
+
+function convert_size_to_fs_blocks(){
+    local device=$1
+    local size=$2
+    block_size_in_bytes=$(/usr/sbin/tune2fs -l "$device" | /usr/bin/awk '/Block size:/{print $3}')
+    echo $(( size / block_size_in_bytes ))
+}
+
+function calculate_expected_resized_file_system_size_in_blocks(){
+    local device=$1
+    increment_boot_partition_in_blocks=$(convert_size_to_fs_blocks "$device" "$INCREMENT_BOOT_PARTITION_SIZE_IN_BYTES")
+    total_block_count=$(/usr/sbin/tune2fs -l "$device" | /usr/bin/awk '/Block count:/{print $3}')
+    new_fs_size_in_blocks=$(( total_block_count - increment_boot_partition_in_blocks ))
+    echo $new_fs_size_in_blocks
+}
+
+function check_filesystem_size() {
+    local device=$1
+    local new_fs_size_in_blocks=$2
+    new_fs_size_in_blocks=$(calculate_expected_resized_file_system_size_in_blocks "$device")
+# it is possible that running this command after resizing it might give an even smaller number. 
+    minimum_blocks_required=$(/usr/sbin/resize2fs -P "$device" 2> /dev/null | /usr/bin/awk  '{print $NF}')
+
+    if [[ "$new_fs_size_in_blocks" -le "0" ]]; then
+        echo "Unable to shrink volume: New size is 0 blocks"
+        return 1
+    fi
+        if [[ $minimum_blocks_required -gt $new_fs_size_in_blocks ]]; then
+        echo "Unable to shrink volume: Estimated minimum size of the file system $1 ($minimum_blocks_required blocks) is greater than the new size $new_fs_size_in_blocks blocks" >&2
+        return 1
+    fi
+    return 0
+}
+
+function process_entry() {
+    is_lvm "$1" "$3"
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        return "$status"
+    fi
+    expected_size_in_bytes=$(parse_tag "$2")
+    if [[ -z "$expected_size_in_bytes" ]]; then
+        echo "Error: Tag $SHRINK_TAG not found for device '$3' in '$2'" >&2
```

Comment:
Let's make this warning rather than error... 
```suggestion
        echo "Warning: Tag $SHRINK_TAG not found for device '$3' in '$2'" >&2
```

Reply:
Done